File: agentwallet/packages/sdk-python/src/agentwallet/resources/swarms.py
"""Agent Swarm sub-resource."""

from __future__ import annotations

from typing import TYPE_CHECKING

from ..types import ListResponse, Swarm, SwarmMember, SwarmTask

if TYPE_CHECKING:
    from ..client import AgentWallet


class SwarmsResource:
    def __init__(self, client: AgentWallet):
        self._client = client

    # ── Swarm CRUD ──

    async def create(
        self,
        name: str,
        description: str,
        orchestrator_agent_id: str,
        swarm_type: str = "general",
        max_members: int = 10,
        is_public: bool = False,
        config: dict | None = None,
    ) -> Swarm:
        body: dict = {
            "name": name,
            "description": description,
            "orchestrator_agent_id": orchestrator_agent_id,
            "swarm_type": swarm_type,
            "max_members": max_members,
            "is_public": is_public,
        }
        if config:
            body["config"] = config
        data = await self._client.post("/swarms", json=body)
        return Swarm(**data)

    async def get(self, swarm_id: str) -> Swarm:
        data = await self._client.get(f"/swarms/{swarm_id}")
        return Swarm(**data)

    async def list(
        self,
        is_public: bool | None = None,
        limit: int = 20,
        offset: int = 0,
    ) -> ListResponse:
        params: dict = {"limit": limit, "offset": offset}
        if is_public is not None:
            params["is_public"] = is_public
        data = await self._client.get("/swarms", params=params)
        return ListResponse(
            data=[Swarm(**s) for s in data["swarms"]],
            total=data["total"],
        )

    # ── Members ──

    async def add_member(
        self,
        swarm_id: str,
        agent_id: str,
        role: str = "worker",
        specialization: str | None = None,
        is_contestable: bool = True,
    ) -> SwarmMember:
        body: dict = {
            "agent_id": agent_id,
            "role": role,
            "is_contestable": is_contestable,
        }
        if specialization:
            body["specialization"] = specialization
        data = await self._client.post(f"/swarms/{swarm_id}/members", json=body)
        return SwarmMember(**data)

    async def list_members(self, swarm_id: str) -> ListResponse:
        data = await self._client.get(f"/swarms/{swarm_id}/members")
        return ListResponse(
            data=[SwarmMember(**m) for m in data["members"]],
            total=data["total"],
        )

    async def remove_member(self, swarm_id: str, agent_id: str) -> dict:
        return await self._client.delete(f"/swarms/{swarm_id}/members/{agent_id}")

    # ── Tasks ──

    async def create_task(
        self,
        swarm_id: str,
        title: str,
        description: str,
        task_type: str = "general",
        client_agent_id: str | None = None,
    ) -> SwarmTask:
        body: dict = {
            "title": title,
            "description": description,
            "task_type": task_type,
        }
        if client_agent_id:
            body["client_agent_id"] = client_agent_id
        data = await self._client.post(f"/swarms/{swarm_id}/tasks", json=body)
        return SwarmTask(**data)

    async def get_task(self, swarm_id: str, task_id: str) -> SwarmTask:
        data = await self._client.get(f"/swarms/{swarm_id}/tasks/{task_id}")
        return SwarmTask(**data)

    async def list_tasks(
        self,
        swarm_id: str,
        status: str | None = None,
        limit: int = 20,
        offset: int = 0,
    ) -> ListResponse:
        params: dict = {"limit": limit, "offset": offset}
        if status:
            params["status"] = status
        data = await self._client.get(f"/swarms/{swarm_id}/tasks", params=params)
        return ListResponse(
            data=[SwarmTask(**t) for t in data["tasks"]],
            total=data["total"],
        )

    async def assign_subtask(
        self,
        swarm_id: str,
        task_id: str,
        subtask_id: str,
        agent_id: str,
        description: str,
    ) -> SwarmTask:
        data = await self._client.post(
            f"/swarms/{swarm_id}/tasks/{task_id}/assign",
            json={
                "subtask_id": subtask_id,
                "agent_id": agent_id,
                "description": description,
            },
        )
        return SwarmTask(**data)

    async def complete_subtask(
        self,
        swarm_id: str,
        task_id: str,
        subtask_id: str,
        result: dict,
    ) -> SwarmTask:
        data = await self._client.post(
            f"/swarms/{swarm_id}/tasks/{task_id}/complete",
            json={"subtask_id": subtask_id, "result": result},
        )
        return SwarmTask(**data)

File: agentwallet/packages/sdk-ts/src/resources/swarms.ts
import type { AgentWallet } from '../client';
import type { CreateSwarmParams, ListResponse, Swarm, SwarmMember, SwarmTask } from '../types';

export class SwarmsResource {
  constructor(private client: AgentWallet) {}

  // ── Swarm CRUD ──

  /** Create a swarm cluster */
  async create(params: CreateSwarmParams): Promise<Swarm> {
    const body: Record<string, unknown> = {
      name: params.name,
      description: params.description,
      orchestrator_agent_id: params.orchestrator_agent_id,
      swarm_type: params.swarm_type ?? 'general',
      max_members: params.max_members ?? 10,
      is_public: params.is_public ?? false,
    };
    if (params.config) body.config = params.config;
    return this.client.post<Swarm>('/swarms', body);
  }

  /** Get a swarm by ID */
  async get(swarmId: string): Promise<Swarm> {
    return this.client.get<Swarm>(`/swarms/${swarmId}`);
  }

  /** List swarms */
  async list(options?: { is_public?: boolean; limit?: number; offset?: number }): Promise<ListResponse<Swarm>> {
    const data = await this.client.get<{ swarms: Swarm[]; total: number }>('/swarms', {
      is_public: options?.is_public,
      limit: options?.limit ?? 20,
      offset: options?.offset ?? 0,
    });
    return { data: data.swarms, total: data.total };
  }

  // ── Members ──

  /** Add a member to a swarm */
  async addMember(swarmId: string, agentId: string, options?: { role?: string; specialization?: string; is_contestable?: boolean }): Promise<SwarmMember> {
    const body: Record<string, unknown> = {
      agent_id: agentId,
      role: options?.role ?? 'worker',
      is_contestable: options?.is_contestable ?? true,
    };
    if (options?.specialization) body.specialization = options.specialization;
    return this.client.post<SwarmMember>(`/swarms/${swarmId}/members`, body);
  }

  /** List swarm members */
  async listMembers(swarmId: string): Promise<ListResponse<SwarmMember>> {
    const data = await this.client.get<{ members: SwarmMember[]; total: number }>(`/swarms/${swarmId}/members`);
    return { data: data.members, total: data.total };
  }

  /** Remove a member from a swarm */
  async removeMember(swarmId: string, agentId: string): Promise<void> {
    await this.client.delete(`/swarms/${swarmId}/members/${agentId}`);
  }

  // ── Tasks ──

  /** Create a task in a swarm */
  async createTask(swarmId: string, title: string, description: string, options?: { task_type?: string; client_agent_id?: string }): Promise<SwarmTask> {
    const body: Record<string, unknown> = {
      title,
      description,
      task_type: options?.task_type ?? 'general',
    };
    if (options?.client_agent_id) body.client_agent_id = options.client_agent_id;
    return this.client.post<SwarmTask>(`/swarms/${swarmId}/tasks`, body);
  }

  /** Get a task by ID */
  async getTask(swarmId: string, taskId: string): Promise<SwarmTask> {
    return this.client.get<SwarmTask>(`/swarms/${swarmId}/tasks/${taskId}`);
  }

  /** List tasks in a swarm */
  async listTasks(swarmId: string, options?: { status?: string; limit?: number; offset?: number }): Promise<ListResponse<SwarmTask>> {
    const data = await this.client.get<{ tasks: SwarmTask[]; total: number }>(`/swarms/${swarmId}/tasks`, {
      status: options?.status,
      limit: options?.limit ?? 20,
      offset: options?.offset ?? 0,
    });
    return { data: data.tasks, total: data.total };
  }

  /** Assign a subtask to a worker agent */
  async assignSubtask(swarmId: string, taskId: string, subtaskId: string, agentId: string, description: string): Promise<SwarmTask> {
    return this.client.post<SwarmTask>(`/swarms/${swarmId}/tasks/${taskId}/assign`, {
      subtask_id: subtaskId,
      agent_id: agentId,
      description,
    });
  }

  /** Complete a subtask with results */
  async completeSubtask(swarmId: string, taskId: string, subtaskId: string, result: Record<string, unknown>): Promise<SwarmTask> {
    return this.client.post<SwarmTask>(`/swarms/${swarmId}/tasks/${taskId}/complete`, {
      subtask_id: subtaskId,
      result,
    });
  }
}
